### PR TITLE
Use default literals for correlated aggregations

### PR DESCRIPTION
**Pull request by ClarenceThreepwood**
Presto implements subqueries with correlated aggregations as an outer join with a subsequent cross join to ensure that the null values produced by the aggregation get coalesced appropriately. In some common cases we are able to substitute the default values directly as a filter. This results in three advantages - we avoid the cross join, the filter could get pushed down and the outer join may then get converted into an inner join (and therefore participate in join reordering) since the filter eliminates nulls. 

This behavior may be controlled through a session parameter "use_defaults_for_correlated_aggregations"

Test plan - (Please fill in how you tested your changes)

Verified results against tpc-h and tpc-ds. 
Test cases in TestUseDefaultsForCorrelatedAggregations and TestPushAggregationThroughOuterJoinWithDefaultsForCorrelatedAggregations

== NO RELEASE NOTE ==

### CODE REVIEW

**Review by bmckennaah — APPROVED** [review]

**Comment by yzhang1991** [comment]
Hi @prestodb/committers, can we have a final review on this?

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:257]: i would give this a different name as its not very clear from a user perspective what this does.  Also, even though the most common case for the aggregation push down is correlated aggregations, the optimizer is generic.  I also wonder if this needs its own session property or if its okay to just guard it with push_aggregation_through_join.  But on the other hand, since bugs in this change could cause wrong results, maybe it is good to have its own flag.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java:268]: I think for this case we should be able to get the FunctionMetadata and check !isCalledOnNullInput().  The only thing is, I'm not sure these are set properly for count and approx_distinct function (i couldn't see that they were set different from the default of false), so that may need to be fixed.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java:265]: I think count(column) over a single null row returns 0, but count(*) returns 1.  did a simple test to try it out.
```
 SELECT count(x) from ( SELECT null as x from dim_one_row);
 _col0                
-------
     0 
(1 row)

SELECT count(*) from ( SELECT null as x from dim_one_row);
 _col0                
-------
     1 
(1 row)
```

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java
Comment (ClarenceThreepwood) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:257]: > i would give this a different name as its not very clear from a user perspective what this does.  

Hard to come up with a concise and descriptive name. How about "optimizer.use-defaults-for-correlated-aggregation-pushdown-through-outer-joins" and I will also do as @mbasmanova suggests and add a more detailed description to the doc

>I also wonder if this needs its own session property or if its okay to just guard it with push_aggregation_through_join. But on the other hand, since bugs in this change could cause wrong results, maybe it is good to have its own flag.

So maybe leave in the flag - but default to true?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java
Comment (ClarenceThreepwood) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java:268]: If I understand correctly, you are suggesting that for all functions for which !calledOnNullInput we should be able to replace with constantNull - not just min, max, sum & avg. 
Makes sense to me . I opened another [PR](https://github.com/prestodb/presto/pull/19123) to set isCalledOnNullInput for aggregate functions to avoid cluttering up this PR. Let's get that merged prior to this

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java
Comment (ClarenceThreepwood) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java:265]: While this is true, I'm not sure it is relevant. At least I haven't been able to come up with a case that breaks the implementation. 
Previously we were adding the cross join in order to coalesce the nulls produced by the outer join to appropriate aggregations over nulls. With this change, the outer join gets converted into an inner anyway (due to the introduction of a null-intolerant predicate) and so cannot produce nulls. The only remaining question is whether the correlated count in the subquery produces a single null row as output - but given that it is a count aggregation (with * or column as input), this can never happen. So I do not think the behavior of count(null) is relevant in this case. Thoughts?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java
Comment (aaneja) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java:268]: > I'm not sure these are set properly for count and approx_distinct function 

isCalledOnNullInput resolves to `false` for these functions. See [this doc](https://docs.google.com/spreadsheets/d/1H3Q11HqK8k9AWOUP_jELISttdPUQsL58SWCvWPyKQxY/edit#gid=375697862) and related [PR comment](https://github.com/prestodb/presto/pull/18938#discussion_r1111558460)

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java
Comment (mbasmanova) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:257]: > Hard to come up with a concise and descriptive name. 

Naming is indeed hard. It may help to write up the documentation for the flag first. Once we have a user-friendly description we may find it easier to come up with a nice name for the flag.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java:268]: yeah, i think we can leave to just this list for now until we fix this field for aggregation functions that need it.  Ideally it would be nice to be able to make this more expansive and check isCalledOnNullInput

Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java:268]: oh just saw your other PR does go through all aggregations to set this correctly.  It's marked as a draft. Is it ready for review now or should I wait?

Comment (ClarenceThreepwood) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java:268]: Just having trouble getting a clean test run

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:257]: defaulting the flag to true seems reasonable to me.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java:265]: can we look specifically for count(value).  i don't want to get wrong results if we start to support count(*)

**Review by rschlussel — COMMENTED** [review]
Can you add correctness tests to AbstractTestQueries.  There might be some coverage already from the correlated subquery tests, but not sure whether it tests all cases (count, null returning function, non-null returning and non-count function)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java
Comment (ClarenceThreepwood) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java:265]: Is this what you were suggesting?

Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushAggregationThroughOuterJoin.java:265]: yes, that should do it!

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java
Comment (rschlussel) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:6417]: you can use a data provider to parametrize the test instead. Then just do regular list of assertQueries.  see example here: https://github.com/prestodb/presto/blob/master/presto-hive/src/test/java/com/facebook/presto/hive/TestHiveLogicalPlanner.java#L827-L850. What do you think?

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java
Comment (rschlussel) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:6424]: I would add comments about what each case tests, eg. max and sum use default null,  approx_distinct, max_by won't use defaults  will do the cross join behavior, count(*) doesn't trigger the optimization.  maybe add a test for count(col) too.

**Review by rschlussel — COMMENTED** [review]
thanks for adding tests!  a couple minor comments.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java
Comment (rschlussel) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:6431]: count(*) doesn't use the pushdown optimization at all.

**Review by rschlussel — APPROVED** [review]

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java
Comment (ClarenceThreepwood) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: Done. 
Fyi - But in this case the optimization will apply to count(*) since it gets rewritten to count(non_null). It will only not kick in when you try to push a count(*) directly through the join - but as you point out earlier, there is another check that prevents this from happening

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java
Comment (rschlussel) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:6431]: never mind. saw your other comment that this gets rewritten to count non-null

**Comment by rschlussel** [comment]
looks like the tpcds plan tests that take advantage of this feature need to be updated to reflect the new plans.

**Comment by rschlussel** [comment]
I suspect this failure in TestSpilledJoinQueries.testInnerJoinWithEmptyBuildSide (https://github.com/prestodb/presto/actions/runs/4369860463/jobs/7644139626) is related to https://github.com/prestodb/presto/pull/19044.  I've asked @feilong-liu to look into it.

**Comment by feilong-liu** [comment]
> I suspect this failure in TestSpilledJoinQueries.testInnerJoinWithEmptyBuildSide (https://github.com/prestodb/presto/actions/runs/4369860463/jobs/7644139626) is related to #19044. I've asked @feilong-liu to look into it.

@rschlussel Submitted a PR for it https://github.com/prestodb/presto/pull/19168

**Pull request merged**
